### PR TITLE
zoom-in-timeseries-drill: do not include :hour and :minute for DATE columns

### LINKED ISSUE #39366: zoom-in-timeseries drill should not include `:hour` and `:minute` for `DATE` columns
Since `DATE` columns don't have time components, we shouldn't try to zoom in to `:hour` or `:minute` on them. `available-drill-thrus`/`zoom-in-timeseries` should not suggest such zooms

### PR DESCRIPTION
**Pull request by appleby**
Closes https://github.com/metabase/metabase/issues/39366

### Description

When the user attempts to drill through a DATE column, we should not return `:hour` or `:minute` drill throughs from `available-drill-thrus`. Previously we did, which resulted in an error being displayed when the user attempted to select them.

Modify `next-breakout-unit` in zoom_in_timeseries.cljc to not return :hour or :minute units for :type/Date columns, which prevents these from being returned by `available-drill-thrus`, which prevents them appearing in UI context menus.

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Databases -> Sample Database -> People
2. Summarize -> Count -> Done
3. Breakout by -> Time -> BIRTH_DATE
4. Select "All Time" by "Day" at the bottom
5. Click on a datapoint in the graph
6. Verify that "See this day by hour" does *not* appear in the context menu

You can also do a similar process for a pivot table visualization, and probably 10 other ways I'm not aware of 😬.

#### before

zoom-in-timeseries-before

before-error

#### after

zoom-in-timeseries-after

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (ranquild) [inline src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: We could use `lib.temporal-bucket/available-temporal-buckets` here. It returns all available options for per column type; but it seems that we need truncation units only. We could use `lib.schema.temporal-bucketing/datetime-truncation-units` for this. Maybe something like:
```
(into []
  (filter (fn [option]
            (->> (lib.temporal-bucket/raw-temporal-bucket option)
                 (contains? lib.schema.temporal-bucketing/datetime-extraction-units))))
  (lib.temporal-bucket/available-temporal-buckets query stage-number column))
```

As a bonus, it:
1. Actually handles `effective-type` correctly
2. Handles custom expressions for which we don't allow temporal buckets in breakouts

Comment (appleby) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: Hi @ranquild, thanks for the review!

I'm glad you mention `lib.temporal-bucket` and `:effective-type` as these were two things I was going to ask about specifically.

In particular, I don't use `available-temporal-buckets` because it is checking `:effictive-type` which turns out to be `:type/Integer` when bucketing by `:year`.

``` clojure
(-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
    (lib/breakout (-> (meta/field-metadata :orders :created-at)
                      (lib/with-temporal-bucket :year)))
    :stages
    first
    :breakout)
==>
[[:field
  {:lib/uuid "d2588602-8018-42d2-92f6-83fb8e7e392b",
   :base-type :type/DateTimeWithLocalTZ,
   :effective-type :type/Integer,
   :temporal-unit :year,
   :metabase.lib.field/original-effective-type :type/DateTimeWithLocalTZ}
  57606]]
```

And, consequently:

``` clojure
(let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
                (lib/aggregate (lib/count))
                (lib/breakout (-> (meta/field-metadata :orders :created-at)
                                  (lib/with-temporal-bucket :year))))
      option {:lib/type :metadata/column,
              :base-type :type/DateTimeWithLocalTZ,
              :effective-type :type/Integer}]
  (lib.temporal-bucket/available-temporal-buckets query -1 option))

==>

[]
```

See, e.g. the comment at the top of

``` clojure
(defmethod lib.temporal-bucket/with-temporal-bucket-method :field
  [[_tag options id-or-name] unit]
  ;; if `unit` is an extraction unit like `:month-of-year`, then the `:effective-type` of the ref changes to
  ;; `:type/Integer` (month of year returns an int). We need to record the ORIGINAL effective type somewhere in case
  ;; we need to refer back to it, e.g. to see what temporal buckets are available if we want to change the unit, or if
  ;; we want to remove it later. We will record this with the key `::original-effective-type`. Note that changing the
  ;; unit multiple times should keep the original first value of `::original-effective-type`.
  (if unit
    (let [extraction-unit?        (contains? lib.schema.temporal-bucketing/datetime-extraction-units unit)
          original-effective-type ((some-fn ::original-effective-type :effective-type :base-type) options)
		  ...
```

The reason being that `lib.schema.temporal-bucketing/datetime-extraction-units` contains `:year`.

That is why I check for `original-effective-type` and `base-type` rather than `effective-type`, but I wanted to ask about that as I was unsure whether that was the right approach...

Also thanks for the tip about filtering `available-temporal-buckets`. I had seen `lib.schema.temporal-bucketing/datetime-truncation-units` but noticed that it was (a) unordered (there is an `ordered-date-truncation-units` but no corresponding `ordered-datetime-truncation-units`, though I guess it would be easy to add) and (b) contains "extra" units like `:second` and `:millisecond` that the old code here didn't include. But I see now that filtering `available-temporal-buckets` by `datetime-truncation-units` seems to be exactly what we'd want, except for the case of `:year` buckets mentioned above.

Comment (ranquild) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: it's usually incorrect to check for either `base-type` or `effective-type` manually. There are many cases to be aware of, like coercion strategies, buckets changing the type etc.

Comment (ranquild) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: are you sure that the behavior you want to achieve with `:year` is correct?

Comment (ranquild) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: @bshepherdson please take a look

**Comment by appleby** [comment]
AFAICT, the test failures look unrelated to this PR. The failing e2e test appears to be flake. The failing bigquery driver test has a name that looks like it could possibly be something broken by this PR, but the actual error appears to just be a read timeout syncing metadata, so maybe not? Not sure what the protocol is for re-running tests when they are suspected flakes...

```
ERROR in metabase.query-processor-test.date-bucketing-test/count-of-grouping-test-2 (get_or_create.clj:145)

:bigquery-cloud-sdk
 4 checkins per hour dataset group by hour
filter by hour = [:relative-datetime :current]
expected: (= 4 (apply count-of-grouping checkins:4-per-hour :hour args))
  actual: 
                 clojure.lang.ExceptionInfo: Failed to sync test database "interval_900": Read timed out
```

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (bshepherdson) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: I think the right way to do this is using `available-temporal-buckets-method` on the breakout ref. That refers to the type of the original field, ignoring any bucketing, which should be `Date`, `DateTime`, or `Time`, and return the right set.

So I'd approach it like this:
- `lib.temporal-bucket/available-temporal-buckets` to get the list for that column
- create `lib.schema.temporal-bucketing/ordered-datetime-truncation-units` since it doesn't exist yet
- filter the `ordered-datetime-truncation-units` to only those which are `available`
- the next one in that list is the one!

Comment (appleby) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: @bshepherdson how do I get the breakout ref? I'm guessing it's something other than the thing I'm selecting in my example above, but I didn't see anything in the query or context that is passed in that looks like a reference to a breakout with the original field's `effective-type`.

In other words, not this, I assume:

```clojure
(-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
    (lib/breakout (-> (meta/field-metadata :orders :created-at)
                      (lib/with-temporal-bucket :year)))
    :stages
    first
    :breakout)
==>
[[:field
  {:lib/uuid "d2588602-8018-42d2-92f6-83fb8e7e392b",
   :base-type :type/DateTimeWithLocalTZ,
   :effective-type :type/Integer,
   :temporal-unit :year,
   :metabase.lib.field/original-effective-type :type/DateTimeWithLocalTZ}
  57606]]
```

Or can I use that uuid to look it up?

Comment (bshepherdson) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: That's the correct ref. The right way to access it is with `(first (lib.breakout/breakouts query -1))`, though. Disregard its `:effective-type`; `available-temporal-buckets` should do the right thing anyway.

(It looks up the original input field rather than relying on the breakout directly, see the `(defmethod available-temporal-buckets-method :field ...)` in `lib.field`.)

Comment (bshepherdson) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: For context, this nonsense is the kind of thing I mean when I said in the retro this morning that refs suck and are the source of much of our pain, bugs and perf waste.

In the prototype I mentioned, breakouts are themselves columns and contain a reference (the ID of another entity in the "database"; which is to say a real, tangible reference rather than a symbolic label!) so you can navigate to the input column with a `get`: `(:mbql.breakout/column a-breakout)` and done.

Comment (appleby) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:53]: Yep, that works! I will rework to use this method instead. Thanks!

```clojure
(let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
                (lib/aggregate (lib/count))
                (lib/breakout (-> (meta/field-metadata :orders :created-at)
                                  (lib/with-temporal-bucket :year))))
      breakout (first (lib.breakout/breakouts query -1))]
  (into []
        (comp 
         (map lib.temporal-bucket/raw-temporal-bucket)
         (filter #(contains? lib.schema.temporal-bucketing/datetime-truncation-units %)))
        (lib.temporal-bucket/available-temporal-buckets query -1 breakout)))

==>

[:minute :hour :day :week :month :quarter :year]
```

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (appleby) [inline src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:76]: Is `:mbql.clause/field` the right type here? I'm assuming that the `(lib.util/clause-of-type? breakout :field)` in `matching-breakout-dimension` means we're guaranteed to get a `:mbql.clause/field` here.

**Comment by appleby** [comment]
@ranquild @bshepherdson

This is ready for re-review whenever you get a chance.

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (ranquild) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:76]: I'd use `column` everywehre

**Review by ranquild — COMMENTED** [review]
Looks good! I'll let @bshepherdson approve. I think `field :- :mbql.clause/field` might be wrong.

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (bshepherdson) [inline src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:58]: This is a lot harder to read than a direct `->>` or `(reverse (into [] (comp (map ...) (filter ...)) ...))`.

Also, note that sets can be used as functions, so there's no need for `contains?` here; just `(filter lib.schema.temporal-bucketing/datetime-truncation-units)`.

You could write this as
```clojure
(reverse (into [] (comp (map lib.temporal-bucket/raw-temporal-bucket)
                        (filter lib.schema.temporal-bucketing/datetime-truncation-units))
                  (lib.temporal-bucket/available-temporal-buckets query stage field)))
```

but I honestly think it'll be better just to use `->>` naively. This list is too short to worry about any overhead.

I'll also point out in passing that `(comp f g)` applies `g`, then `f`, ie. it's right-to-left. **But** transducers are functions that wrap an abstract pipeline with a new, pre-order transformation. So `(comp (map f) (filter g))` produces a transducer that maps `f` and then filters with `g`.

```clojure
(->> (lib.temporal-bucket/available-temporal-buckets query stage field)
     (map lib.temporal-bucket/raw-temporal-bucket)
     (filter lib.schema.temporal-bucketing/datetime-truncation-units)
     reverse)
```

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (bshepherdson) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:76]: Yes, I think that's correct here.

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (bshepherdson) [inline src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:79]:
```suggestion
    (->> (valid-current-units query stage field)
         (drop-while #(not= % current-unit))
         second)))
```

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (appleby) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:58]: Awesome, thanks for the tip. As you can tell I'm still struggling with idiomatic Clojure, so these tips are appreciated. Will update the PR with these suggestions.

**Comment by appleby** [comment]
Rebased onto master and updated the label from `no-backport` to `backport`

File: test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc
Comment (camsaul) [inline test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc:166]: A lot of our tests end in `test-n` or whatever as a byproduct of breaking out giant tests into smaller ones, it's even better to give different tests more meaningful names if you can. Something like `returns-zoom-in-timeseries-count-date-by-month-test` or something and then `-sum-date-by-month-test` for the next one is a little bit better than just giving every test a number .

I would really consider combining these all into one test tho, maybe something like

```clj
(deftest ^:parallel returns-zoom-in-timeseries-date-column-test
  (testing "zoom-in.timeseries should be returned for Date dimension (#33811, #39366)"
    (doseq [aggregation ["count" "sum" "max"]
            bucketing   [:month :day]]
      (testing (lib.util/format "aggregation = %s, bucketing = %s" aggregation bucketing)
        (lib.drill-thru.tu/test-returns-drill
         (zoom-in-timeseries-drill-for-orders-created-at
          metadata-provider-with-orders-created-at-as-date
          aggregation
          bucketing))))))
```

otherwise it gets really repetitive and noisy to write out 6 versions of the same test that only differ by one or two things

**Review by camsaul — APPROVED** [review]
Well overall this LGTM but I would consider combining all those very similar tests into one single test with `doseq` to make things a little more concise before merging

**Comment by appleby** [comment]
> Well overall this LGTM but I would consider combining all those very similar tests into one single test with `doseq` to make things a little more concise before merging

That's actually how I implemented it in a first pass, then noticed the existing tests for DateTime were separate tests and copied that, thinking maybe we prefer to break them out to increase parallelism or else just prefer to have separate test cases to make it easier to identify which test failed.

Will update before merging.

File: src/metabase/lib/drill_thru/zoom_in_timeseries.cljc
Comment (appleby) [reply src/metabase/lib/drill_thru/zoom_in_timeseries.cljc]: Fixed

File: test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc
Comment (appleby) [reply test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc]: Fixed in latest commit

**Pull request merged**
